### PR TITLE
runtime(compiler): Introduce event-driven primitives for SpotBugs

### PR DESCRIPTION
**Pull request by zzzyxwvut**

### CODE REVIEW

**Comment by zzzyxwvut** [comment]
@Konfekt, an asynchronous, [follow-up](https://github.com/vim/vim/pull/16140#issuecomment-2532756748) example with [`:Make`](https://github.com/tpope/vim-dispatch/blob/a2ff28abdb2d89725192db5b8562977d392a4d3f/doc/dispatch.txt#L15) from  
`vim-dispatch`:

```vim
" ASSUME THAT:
" * ":Make" is capable of asynchronous execution and consuming ShellCmdPost
"   events.
" * The default implementations for "spotbugs#DefaultPreCompiler*Action" and
"   "spotbugs#DefaultPostCompilerAction" are functional; otherwise, define
"   "Action"s proper not "Command"s.
" * Neither "PreCompiler*Action" nor "PostCompilerAction" is defined, or if
"   defined, each implementation allows for "Command" re-use (see examples in
"   "autoload/spotbugs.vim").

if exists('g:spotbugs_properties') && exists(':Make') == 2
	" Request an asynchronous strategy.
"	let g:dispatch_no_tmux_make = 1
"	let g:dispatch_no_screen_make = 1

	set cmdheight=16

	function! VimDispatchPreCompilerCommand(arguments) abort
		if !exists('g:spotbugs_compilation_done')
			doautocmd java_spotbugs_post User
			execute 'Make ' . a:arguments
			let g:spotbugs_compilation_done = 1
		else
			" Follow a synchronous path in the absence of class
			" file compilation.
			cc
		endif
	endfunction

	function! VimDispatchPreCompilerTestCommand(arguments) abort
		if !exists('g:spotbugs_test_compilation_done')
			doautocmd java_spotbugs_post User
			execute 'Make ' . a:arguments
			let g:spotbugs_test_compilation_done = 1
		else
			" Follow a synchronous path in the absence of class
			" file compilation.
			cc
		endif
	endfunction

	function! VimDispatchPostCompilerCommand(arguments) abort
		execute 'Make ' . a:arguments
	endfunction

	function! GenericPostCompilerActionExecutor(action) abort
		try
			cc
		catch /\<E42:/
			execute a:action
		endtry
	endfunction

	let g:spotbugs_properties.DefaultPreCompilerCommand =
		\ function('VimDispatchPreCompilerCommand')
	let g:spotbugs_properties.DefaultPreCompilerTestCommand =
		\ function('VimDispatchPreCompilerTestCommand')
	let g:spotbugs_properties.DefaultPostCompilerCommand =
		\ function('VimDispatchPostCompilerCommand')
	let g:spotbugs_properties.PostCompilerActionExecutor =
		\ function('GenericPostCompilerActionExecutor')
endif
```
See the updated documentation [entry](https://github.com/vim/vim/blob/3aabd151bf3d29c643d5a92cc4fec521130a40db/runtime/doc/quickfix.txt#L1427-L1484).

**Comment by Konfekt** [comment]
Thank you!

Since `:Make` is the canonical choice to replicate `:make`, couldn't this be detected and used automatically instead of

```
f a command equivalent of `:make` is capable of asynchronous execution and
consuming `ShellCmdPost` events, `:doautocmd java_spotbugs_post ShellCmdPost`
must be removed from such "*Action" (or "*Command") implementations (i.e. the
lines `(a)` and `(b)` in the listed examples) to retain a sequential order for
a non-blocking execution, and any notification (see below) must be suppressed.
```

leaving this removal up to the user?

**Comment by zzzyxwvut** [comment]
With the proposed changes, a batch of primitives is offered  
for copying post-compiler actions (or their commands) into  
a once-only `ShellCmdPost` command and delegating to user  
implementations the control over their execution:

1. `PostCompilerActionExecutor`; 
2. `doautocmd java_spotbugs_post User`;
3. `doautocmd java_spotbugs_post ShellCmdPost`.

Some use of primitives is suited for synchronous execution  
(1, 2, 3); another use is suited for asynchronous execution  
(1, 2).  As before, it is still in the hands of users to  
obtain and configure custom `:Make`s.

The documented example shows what definitions and the kind  
of command arrangement is necessary to set up _synchronous_  
event-driven execution for a Maven project with `:make` (the  
builtin command).  The quoted excerpt from the documentation  
entry explains how to modify that example in order to enable  
_asynchronous_ event-driven execution for an unspecified  
equivalent of `:make`, assuming the listed requirements for  
it are met.

The above snippet is an example setup for _asynchronous_   
event-driven execution that can be used in general (no Maven  
attachment) with `:Make` available from the `vim-dispatch`  
plugin.  (Its linked counterpart snippet is an example setup  
for _synchronous non-event-driven_ execution.)  It can be put  
in some `~/.vim/plugin/spotbugs.vim` and, after removing the  
assignment statements at the bottom, its Funcrefs can be  
considered for any `g:spotbugs_properties` on a case-by-case  
basis (with attendant checks for `:Make`'s availability).

So, conditionally-defined primitives and examples with these  
primitives are submitted; default implementations are not  
touched.

File: runtime/doc/quickfix.txt [hunk 1]
```diff
@@ -1424,6 +1424,66 @@ of |:make|, and assigning its |Funcref| to the selected key.  For example:
 			\ function('GenericPostCompilerCommand'),
 	\ }
 
+When "PostCompilerAction" is available, "PostCompilerActionExecutor" is also
+supported.  Its value must be a Funcref pointing to a function that always
+declares an only parameter of type string and decides whether |:execute| can
```
Comment (chrisbra) [inline runtime/doc/quickfix.txt:1429]:
```suggestion
declares a single parameter of type string and decides whether |:execute| can
```

**Comment by chrisbra** [comment]
Thanks. I include it + my suggestion :) And I emphasize the Note `XXX` a bit more

**Comment by zzzyxwvut** [comment]
The added comments are helpful.  Thank you.

**Pull request closed without merging**